charmcraft/main.py

```diff
@@ -192,32 +198,56 @@ def _pre_parse_args(self, sysargs):
 
         - validate that command is correct (NOT loading and parsing its arguments)
         """
-        # get all flags (default to False) and those per options, to filter sysargs
-        flags = {}
-        flag_per_option = {}
-        for flag, short_option, long_option, _ in GLOBAL_FLAGS:
-            flag_per_option[short_option] = flag
-            flag_per_option[long_option] = flag
-            flags[flag] = False
+        # get all arguments (default to what's specified) and those per options, to filter sysargs
+        global_args = {}
+        arg_per_option = {}
+        options_with_equal = []
+        for arg in GLOBAL_ARGS:
+            arg_per_option[arg.short_option] = arg
+            arg_per_option[arg.long_option] = arg
+            if arg.type == 'flag':
+                default = False
+            elif arg.type == 'option':
+                default = None
+                options_with_equal.append(arg.long_option + '=')
+            else:
+                raise ValueError("Bad GLOBAL_ARGS structure.")
+            global_args[arg.name] = default
 
         filtered_sysargs = []
-        for arg in sysargs:
-            if arg in flag_per_option:
-                flags[flag_per_option[arg]] = True
+        sysargs = iter(sysargs)
+        options_with_equal = tuple(options_with_equal)
+        for sysarg in sysargs:
+            if sysarg in arg_per_option:
+                arg = arg_per_option[sysarg]
+                if arg.type == 'flag':
+                    value = True
+                else:
+                    try:
+                        value = next(sysargs)
+                    except StopIteration:
+                        raise CommandError("The 'from' option expected one argument.")
```

Comment:
Maybe "expects" instead of "expected"? Or "requires an argument"?

Reply:
much better, thanks!